### PR TITLE
Remove external dependencies from Scyther

### PR DESCRIPTION
**Pull request by stfnhdr**
Hey, awesome Project you made here - switched from netfox to scyther and have big hopes for it.

SnapKit and SDWebImage is cool and all, but i think it's better if a debug-tool does not introduce new dependencies to a project.

This all happened in a quick late night session and a hope i did not f**k things up ^^

### CODE REVIEW

**Comment by stfnhdr** [comment]
i just noticed something is broken on sub iOS13 devices, even before my changes. I'm on it.

Simulator Screen Shot - iPhone 5s - 2021-06-02 at 09 08 49

**Comment by bstillitano** [comment]
> i just noticed something is broken on sub iOS13 devices, even before my changes. I'm on it.
> 
> 
> 
> Simulator Screen Shot - iPhone 5s - 2021-06-02 at 09 08 49
> 
> 

It looks as though you don't have an app icon set in your Assets folder. Is that the case? That would explain the broken label layout.

File: Sources/Scyther/Extensions/UIImageView+Extension.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//
+//  Created by Stefan Haider on 01.06.21.
+//
+
+#if !os(macOS)
+import UIKit
+
+extension UIImageView {
+    
+    func loadImage(fromURL url: URL, completion: ((Bool) -> Void)? = nil) {
```
Comment (bstillitano) [inline Sources/Scyther/Extensions/UIImageView+Extension.swift:13]: Please document this function.

File: Sources/Scyther/Extensions/UIColor+Extensions.swift [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //
-//  File.swift
+//  UIColor+Extension.swift
```
Comment (bstillitano) [inline Sources/Scyther/Extensions/UIColor+Extensions.swift:2]: Missing the `s` on the end of `Extensions`

File: Sources/Scyther/Components/Cells/SliderCell.swift [hunk 3]
```diff
@@ -32,23 +32,29 @@ class SliderCell: UITableViewCell {
     }
     
     private func setupConstraints() {
-        textLabel?.snp.remakeConstraints({ (make) in
-            make.top.equalToSuperview().inset(16)
-            make.left.equalToSuperview().inset(16)
-            make.right.equalToSuperview().inset(16)
-        })
+        guard let textLabel = textLabel else { return }
```
Comment (bstillitano) [inline Sources/Scyther/Components/Cells/SliderCell.swift:35]: Any reason for unwrapping here? Preference is to leave the optional in place.

**Comment by stfnhdr** [comment]
sorry i was so stupid and uploaded a screenshot with the bundle identifier, would be kind of you to also delete your comment, sorry -.-

File: Sources/Scyther/Components/Cells/PreviewableCell.swift [hunk 4]
```diff
@@ -30,36 +30,44 @@ class PreviewableCell: UITableViewCell {
...
-        previewView.snp.remakeConstraints { (make) in
-            make.top.equalTo(descriptionLabel.snp.bottom).offset(16)
-            make.left.bottom.right.equalToSuperview()
-        }
+        NSLayoutConstraint.activate([
+            previewView.topAnchor.constraint(equalTo: descriptionLabel.topAnchor),
+            previewView.leftAnchor.constraint(equalTo: contentView.leftAnchor),
+            previewView.rightAnchor.constraint(equalTo: contentView.rightAnchor),
+            previewView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor)
+        ])
```
Comment (bstillitano) [inline Sources/Scyther/Components/Cells/PreviewableCell.swift:70]: This does not keep the same constraints in place.

**Comment by bstillitano** [comment]
I will review this PR in detail and provide feedback. I am all for removing SDWebImage given its extremely minimal usage, however remain unconvinced on the value of removing SnapKit. 

Regarding introducing new package dependencies, I feel as though this is quite common place and is exactly what SPM was designed to handle. There is definitely something to be said about introducing as few dependencies as possible but I will need to undertake extensive testing before we merge this in and remove SnapKit.

File: Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -631,20 +631,20 @@
 /* End XCConfigurationList section */
 
 /* Begin XCRemoteSwiftPackageReference section */
-		CA73FDDB25DA22E8007FA627 /* XCRemoteSwiftPackageReference "Scyther" */ = {
+		C9A3A2362666BAD400E2B2F9 /* XCRemoteSwiftPackageReference "Scyther" */ = {
 			isa = XCRemoteSwiftPackageReference;
-			repositoryURL = "https://github.com/bstillitano/Scyther.git";
+			repositoryURL = "file:///Users/stefanhaider/Documents/XcodeWorkespace/Scyther";
```
Comment (bstillitano) [inline Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj:636]: Please revert this change.

File: Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -631,20 +631,20 @@
...
 			requirement = {
-				branch = main;
+				branch = stfnhdr_remove_depenencies;
```
Comment (bstillitano) [inline Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj:638]: Please revert this change.

File: Scyther Playground/Scyther Playground.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 7]
```diff
@@ -1,34 +0,0 @@
...
-        "state": {
-          "branch": null,
-          "revision": "d458564516e5676af9c70b4f4b2a9178294f1bc6",
-          "version": "5.0.1"
-        }
-      }
-    ]
-  },
-  "version": 1
-}
```
Comment (bstillitano) [inline Scyther Playground/Scyther Playground.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:34]: Why is this file being deleted? This will break the repository.

File: Scyther Playground/Scyther Playground/Button.swift [hunk 8]
```diff
@@ -170,12 +170,17 @@ class Button: UIButton {
...
     }
 
     private func setType(_ type: ButtonStyle) {
+        guard let titleLabel = titleLabel,
+              let imageView = imageView,
+              let superview = superview
+        else {
+            return
+        }
+
```
Comment (bstillitano) [inline Scyther Playground/Scyther Playground/Button.swift:183]: Again, preference is to not unwrap unnecessarily. Nothing in this file requires that these are explicitly not null.

File: Scyther Playground/Scyther Playground/Button.swift [hunk 9]
```diff
@@ -187,23 +192,25 @@ class Button: UIButton {
...
         setTitleColor(type.disabledBackgroundColor, for: .disabled)
         
         //Set Constraints
-        titleLabel?.snp.remakeConstraints({ (make) in
-            make.height.equalTo(type.buttonHeight)
-            make.top.bottom.equalToSuperview()
-        })
+        NSLayoutConstraint.activate([
+            titleLabel.heightAnchor.constraint(equalToConstant: type.buttonHeight),
+            titleLabel.topAnchor.constraint(equalTo: superview.topAnchor),
```
Comment (bstillitano) [inline Scyther Playground/Scyther Playground/Button.swift:211]: Why is there now a new top anchor here?

File: Sources/Scyther/Components/Cells/DeviceCell.swift [hunk 10]
```diff
@@ -9,41 +9,53 @@
...
-            make.right.equalToSuperview()
-        })
+        
+        guard let imageView = imageView,
+              let textLabel = textLabel,
+              let detailTextLabel = detailTextLabel
+        else {
+            return
+        }
+
```
Comment (bstillitano) [inline Sources/Scyther/Components/Cells/DeviceCell.swift:22]: Same thing with the unwrapping here. Not required.

**Review by bstillitano — CHANGES_REQUESTED** [review]
There are multiple concerns here that would require addressing before we can merge this. 

Are you able to address these and I can give it another review?

File: Sources/Scyther/Components/Cells/SliderCell.swift (hunk 3 above)
Comment (stfnhdr) [reply Sources/Scyther/Components/Cells/SliderCell.swift:35]: the classic way to set layoutconstraints unfortunately does not work with optionals, but i can do it like this if preferred:

```
        if let textLabel = textLabel {
            NSLayoutConstraint.activate([
                textLabel.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 16),
                textLabel.leftAnchor.constraint(equalTo: contentView.leftAnchor, constant: 16),
                textLabel.rightAnchor.constraint(equalTo: contentView.rightAnchor, constant: -16)
            ])
        }
```

File: Scyther Playground/Scyther Playground/Button.swift (hunk 8 above)
Comment (stfnhdr) [reply Scyther Playground/Scyther Playground/Button.swift:183]: same as above i could solve it like this if preferred:

```
        if let textLabel = textLabel {
            NSLayoutConstraint.activate([
                textLabel.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 16),
                textLabel.leftAnchor.constraint(equalTo: contentView.leftAnchor, constant: 16),
                textLabel.rightAnchor.constraint(equalTo: contentView.rightAnchor, constant: -16)
            ])
        }
```

File: Scyther Playground/Scyther Playground/Button.swift (hunk 9 above)
Comment (stfnhdr) [reply Scyther Playground/Scyther Playground/Button.swift:211]: the old code had a hight constraint and a top and bottom constraint:
```
titleLabel?.snp.remakeConstraints({ (make) in
            make.height.equalTo(type.buttonHeight)
            make.top.bottom.equalToSuperview()
        })
```

File: Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj (hunk 5 above)
Comment (stfnhdr) [reply Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj:636]: i already changed that the playground uses the local swift package so someone can play around locally and experiment a little, if this is a problem i'll change it back to the original implementation.

File: Sources/Scyther/Components/Cells/SliderCell.swift (hunk 3 above)
Comment (stfnhdr) [reply Sources/Scyther/Components/Cells/SliderCell.swift:35]: but in this case, cause all the constraint refer to the textLabel i think its cleaner to catch this with a precondition.

File: Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj (hunk 5 above)
Comment (bstillitano) [reply Scyther Playground/Scyther Playground.xcodeproj/project.pbxproj:636]: The playground is now pointing to your local folder on your machine. This will not work on anyone else's machine but yours.

File: Sources/Scyther/Extensions/UIImageView+Extension.swift [hunk 11]
```diff
@@ -0,0 +1,26 @@
+//
+//  UIImageView+Extension.swift
```
Comment (bstillitano) [inline Sources/Scyther/Extensions/UIImageView+Extension.swift:2]: This filename should be pluralised `UIImageView+Extensions.swift`

File: Sources/Scyther/Components/Cells/SliderCell.swift (hunk 3 above)
Comment (bstillitano) [reply Sources/Scyther/Components/Cells/SliderCell.swift:35]: That doesn't sound right. I think the issue is that you're trying to do it via the array invocation `NSLayoutConstraint.activate([])`. You could do

```
textLabel?.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 16).isActive = true,
textLabel?.leftAnchor.constraint(equalTo: contentView.leftAnchor, constant: 16).isActive = true,
textLabel?.rightAnchor.constraint(equalTo: contentView.rightAnchor, constant: -16).isActive = true
```

I still remain unconvinced of the value that this change brings. At this point the complexity that the changes would introduce to the code seem like they negate any benefits that these changes would bring.

**Comment by bstillitano** [comment]
@stfnhdr Are you able to please separate this PR into two different ones.

One that removes SDWebImage and one that removes SnapKit. I ask as I do not want to completely negate your contribution however I do feel as though there are not compelling enough arguments right now that convince me that SnapKit should be removed. Would like to merge your SDWebImage stuff and then continue the conversation over on a separate PR for the SnapKit stuff.

File: Sources/Scyther/Components/Cells/SliderCell.swift (hunk 3 above)
Comment (stfnhdr) [reply Sources/Scyther/Components/Cells/SliderCell.swift:35]: Good point - if you prefer it this way i will change that

**Comment by stfnhdr** [comment]
Sounds good, yeah let's close this PR and make two separate PRs. I'll fix some error and make it a not late night action ;)

**Comment by stfnhdr** [comment]
Sounds good, yeah let's close this PR and make two separate PRs. I'll fix some error and make it a not late night action ;)

**Comment by bstillitano** [comment]
Awesome. Thanks for understanding 😄
